utils/restendpointhelper.go

```diff
@@ -0,0 +1,67 @@
+package utils
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+const restUrlPatternHeaderKey string = "Owncast-Resturl-Pattern"
```

Comment:
I'm a little confused about this.  It seems like you're setting a header value as a guard, so only requests with that header is allowed to use the path parameters.  But it feels kind of weird.  Can we just return no parameters if none exist?

Reply:
Oh and as just a random aside, you don't need to specify the type `string` if you're assigning a value, but if you're declaring an empty var then you do.

Reply:
Okay, I feel like I should have outlined how the REST-endpoint magic works (or how its supposed to work).

Initially I was investigating the `gorilla/mux` framework for matching REST endpoints. As discussed in a previous thread, there is and will be the need for REST endpoint matching, but we should stay away from third-party dependencies if possible and rather implement our own. So I implemented a very very simple way to handle such endpoints.

First I want to explain how I intended to parse Variables from the Path/Segment of an Url:
suppose we have a URL-segment that looks like this `/user/12/friend/13/name` and we want to parse out `12` and `13`.  A simple (but very unsafe way) to parse these out would be to split the string by '/' and access the vars by accessing split[0] and split[2]. The more reliable and secure version, is to match the Url against a pattern. So this is where this `Rest-Url-Pattern` comes into play. Basically we define a pattern (other web-frameworks like Flask, ASP.NET modelbinder, refer to this as "Segment-Mask") that is used as Pattern-String for the Input URL. Given the above example, the mask would look like this: `/user/{userId}/friend/{friendId}/name`. That mask helps to safely extract the actual parameters out of the URL path. 

#### Extracting an parameter out of the URL path is implemented like this

* need the REST-pattern and the URL-segment (current URL Request Path) as string 
* Split both the REST-pattern and the actual URL-segment by `/`
* the resulting arrays/slices  must have equal length/dimensions (if not, something is wrong, abort)
* to lookup the location of the `{clientId}` parameter in the Url:
    *  find the Index in the split of the REST-pattern (the index that contains the placeholder `{clientId}`)
    *  and then use this index to access the element at the same index in the URL-Segment split array.
 * evoila we've safely extracted the input parameter without relying on any assumptions, and without relying more than necessary on user-crafted input

#### You might ask, why this REST-pattern is there at all

* it's a very performant way to parse parameters
* it makes these requests more readable and easier to structure in the mux
* it's safer than directly parsing arbitrary user-input (reducing attack-surface for parameter tampering)
* you can identify faults/errors in the URL-segment immediately (e.g. different dimensions)
* you could additional type-checks something like: `/user/{clientId: int}/friend/{friendId: int}/name` (at one point in the future)

#### Okay, if you're cool with that part so far - let's close the loop

The [`RestEndpoint`- Request-Middleware](https://github.com/owncast/owncast/pull/1351/files#diff-a2a44c38aa0276fd98d206598799c9eec7f0d0700704975ff05ec1f32775168dR61) takes this REST-Pattern and registers a HandlerFunc for this Pattern. Internally this middleware will reduce `/user/{clientId: int}/friend/{friendId: int}/name` to `/user/` and register the passed HandlerFunc with `/user/` to the `net/http` mux.  I had to do it like this, because the vanilla `net/http` server does not support partial Url matching (or placeholders in Urls). 

Inside the Controller-Function in [`controllers/admin/chat.go`](https://github.com/owncast/owncast/blob/5fc62eefdc39f9374911ca373d8b4df05a52f505/controllers/admin/chat.go#L131) I'm using the `clientIdText, err := utils.ReadRestUrlParameter(r, "clientId")` function to parse out the ClientId from the template. Now this function requires the REST-pattern to be present to safely work (see REST-parameter parsing part above). The solution that I came up with, is to pass the REST-Pattern from the `RestEndpoint`-middleware further down the middleware chain until it finally reaches a handler that needs the pattern for parsing out a parameter. The nicest way I could think of, was utilizing the HTTP headers. Now the middleware sets an HTTP Header var with the Pattern, that then can be read from the `utils.ReadRestUrlParameter(r, "clientId")` inside the actual HandlerFunc.

That guard clause (that I think you are referring to) inside the `ReadRestUrlParameter`-func is actually just a guard making sure, that the Rest-Pattern is set by the middleware (or in other words, the developer actually decorated the HandlerFunc with the `RestEndpoint` middleware). In the real world, this guard clause won't ever cause a problem. I see it more or less as a "fail-fast"-shortcircuit for developers using this helpers.

Excuse me for slashing this together like this, also excuse me for the wall of text. I hope this explanation at least somewhat makes sense and is acceptable for you.

Reply:
Thank you for all the detailed clarification!

I went ahead and pushed a few tweaks (mostly renames, comments) needed by the linter and resolved some merge conflicts.